python/dgl/heterograph.py

```diff
@@ -4476,6 +4347,32 @@ def cpu(self):
         """
         return self.to(F.cpu())
 
+    def clone(self):
+        """Return a heterograph object that is a clone of current graph.
+
+        Returns
+        -------
+        DGLHeteroGraph
+            The graph object that is a clone of current graph.
+        """
+        meta_edges = []
+        for s_ntype, e_type, d_ntype in self.canonical_etypes:
+            meta_edges.append((self.get_ntype_id(s_ntype), self.get_ntype_id(d_ntype)))
+
+        metagraph = graph_index.from_edge_list(meta_edges, True)
+        # rebuild graph idx
+        num_nodes_per_type = [self.number_of_nodes(c_ntype) for c_ntype in self.ntypes]
+        relation_graphs = [self._graph.get_relation_graph(self.get_etype_id(c_etype))
+                           for c_etype in self.canonical_etypes]
+        hgidx = heterograph_index.create_heterograph_from_relations(
+            metagraph, relation_graphs, utils.toindex(num_nodes_per_type, "int64"))
```

Comment:
It looks like we could just pass the same graphindex object to the result graph because those inplace mutation operators is actually *out-place* because a new graphindex object is created, which means it is safe to share graph index across different graphs.

Reply:
The graph index in C++ has lots of shared_ptr objects. Will this cause problems?